ktor-client/ktor-client-curl/src/io/ktor/client/engine/curl/CurlState.kt

```diff
@@ -0,0 +1,172 @@
+package io.ktor.client.engine.curl
+
+// This file knows nothing about ktor. Let's keep ti this way.
+import kotlinx.cinterop.*
+import kotlin.native.concurrent.*
+import platform.posix.*
+import libcurl.*
+import libcurl.CURLMsg.*
+
+// Those two are passed between the worker and the driver.
+
+internal class CurlRequestData(
+    val url: String,
+    val method: String,
+    val headers: CPointer<curl_slist>?,
+    val content: ByteArray?
+    // val attributes: Attributes
+)
+
+internal class CurlResponseData(
+    val request: CurlRequestData,
+    val chunks: MutableList<ByteArray> = mutableListOf(),
+    val headers: MutableMap<String, MutableList<String>> = mutableMapOf()
+) {
+    /* lateinit */ var status: Int = 0
+    /* lateinit */ var version: UInt = 0u
+}
+
+// The whole this class is executed in the worker.
+
+internal class CurlState(val existingMultiHandle: COpaquePointer? = null) {
+    val multi_socket = false
+
+    private val multi_handle = existingMultiHandle ?: setupMultiHandle()
+
+    fun setupMultiHandle(): COpaquePointer? {
+        val handle = curl_multi_init() ?: error("Could not initialilze libcurl multi handle")
+
+        return handle
+    }
+
+    fun close() {
+        // Make sure all easy handles have been removed from the multi handle
+        curl_multi_cleanup(multi_handle)
+            .code()
+    }
+
+    fun setupUploadContent(easyHandle: COpaquePointer, content: ByteArray?) {
+        if (content == null) return
+
+        val stream = SimpleByteArrayStream(content)
+
+        // The stable ref is disposed by the callback.
+        easyHandle.apply {
+            option(CURLOPT_READDATA, StableRef.create(stream).asCPointer())
```

Comment:
Probably we should extract some `StableRef.create(stream).asCPointer()` and etc to utility methods.
It looks like we'tr writing a lot of boilerplate code everywhere to use C API.